content/blog/2020-11-04-unified-theory-gc/index.md

```diff
@@ -0,0 +1,176 @@
+  check if a node's reference count is artificially high due to circular
+  references; evidently the case we care about is when a reference count is
+  non-zero purely due to circular references, in which case its memory should be
+  reclaimed).
+
+* **CD: Deferred Reference Counting.** Do reference counting, but save overhead
+  by not counting mutations of root references. Add items
+  to a Zero Count Table (ZCT) when they are decremented to zero. References from the
+  "root set" (a concept borrowed from tracing meaning "references into the heap
+  from outside of the heap") can keep ZCT entries alive.
+
+* **CT: Reference Counting, Tracing Backup.** Use reference counting without
+  attempting to detect cycles. Occasionally do a full tracing collection to
+  clean up cyclic garbage.
+
+* **C: Reference Counting.** This is the classic reference counting algorithm. 
+  They give a really slick way to reduce space
+  overhead for traversing the links by using the space guaranteed to exist in
+  the dead objects themselves.
+
+The paper concludes with a mathematical analysis of the costs of the
+above collection strategies. The authors point out that the analyses are not 
+big-O, so these costs are actually "real", up to the accuracy of the constants
+for a particular system (and I guess also the constancy of those constants?).
+
+The analyses focus on computing the formulas for: 
+* $\phi$, the frequency of garbage collection
+* $\kappa$, the cost of an individual collection, and
+* $\tau$, the total cost of collection, which is generally $\phi \cdot \kappa$
+  plus the cost of mutation overhead
+
+The computation of $\phi$ and $\kappa$ are kept separate, presumably because
+they quanitify the essence of the classic tradeoffs between tracing (large
+pauses) and reference counting (continuous overhead). Thus, by investigating the
+$\phi$ and $\kappa$ values of an approach one might be considering on the system
+one has in mind, one could evaluate where along the spectrum of hybrids the
+approach lies (on that system).
+
+
+## Merits of the paper
+
+The duality between tracing and reference counting itself is a both mind-bending
+and extremely well developed.  Given that the paper is really just offering this
+insight about a nice way to categorize existing garbage collection methods, its
+organization is quite nice.  That is, they give the deep comparison abstractly
+in several different ways, then completely flesh out how existing hybrids
+combine the two flavors.
+
+The section of the paper that has the most potentional to be practically
+important is the cost analysis in section 7.  Having never constructed a garbage
+collection system for a whole language, I do not know whether a
+compiler designer would actually use these formulas to tune a garbage collector
+(even with knowledge of their particular domain) since they would have to
+validate any decisions against benchmark performance anyway.
```

Comment:
Great question; I don't know either! I can see it perhaps being useful as a first step, in combination with validation on real benchmarks.